File: Prototype/BigGp/src/Model/SmallNodes/Expressions/Standart/Childs/MultiplicativeExpression.java
package Model.SmallNodes.Expressions.Standart.Childs;

import Model.Interfaces.Crossable;
import Model.Interfaces.SubtreeMutable;
import Model.Node;
import Model.SmallNodes.Expressions.Standart.Expression;
import Model.SmallNodes.Expressions.Standart.TwoArgExpression;
import Model.SmallNodes.Operators.Arithmetic.MultiplicativeOperator;
import Serialization.Token;

import java.util.ArrayList;

public class MultiplicativeExpression extends Node implements Crossable, TwoArgExpression, SubtreeMutable {

    private Expression leftExpression;
    private MultiplicativeOperator operator;
    private Expression rightExpression;

    @Override
    public String toString() {
        return "(" + leftExpression.toString() + operator.toString() + rightExpression.toString() + ")";
    }

    public void setExpression(Node oldExpr, Expression expression) {
        if (leftExpression.toString().equals(oldExpr.toString()) && leftExpression.hashCode() == oldExpr.hashCode()) {
            this.leftExpression = expression;
        } else {
            this.rightExpression = expression;
        }
    }

    public void subtreeMutate() {
        if (this.getParentNode() instanceof Expression expression) {
            expression.subtreeMutateTwoArgExpression();
        }
    }

    public Expression getLeftExpression(){
        return leftExpression;
    }

    public Expression getRightExpression(){
        return rightExpression;
    }

    public String getOperatorValue(){
        return operator.getValue();
    }

    public ArrayList<Node> getChildrenAsNodes(){
        ArrayList<Node> nodes = new ArrayList<>();
        nodes.add(this);

        nodes.addAll(leftExpression.getChildrenAsNodes());
        nodes.addAll(operator.getChildrenAsNodes());
        nodes.addAll(rightExpression.getChildrenAsNodes());

        return nodes;
    }

    public MultiplicativeExpression(Node parentNode) {
        super(parentNode, "MultiplicativeExpression", true);

        double lastNextTwoArgExpressionChance = ((Expression) parentNode).getNEXT_TWO_ARG_EXPRESSION_CHANCE();
        leftExpression = new Expression(this, lastNextTwoArgExpressionChance);
        operator = new MultiplicativeOperator(this);
        rightExpression = new Expression(this, lastNextTwoArgExpressionChance);
    }

    public MultiplicativeExpression(Node parentNode, Expression left, Expression right, String operator) {
        super(parentNode, "AdditiveExpression", true);

        leftExpression = left;
        this.operator = new MultiplicativeOperator(this);

        while (this.operator.getValue().equals(operator)){
            this.operator = new MultiplicativeOperator(this);
        }
        rightExpression = right;
    }

    public MultiplicativeExpression(Node parentNode, ArrayList<Token> tokens) {
        super(parentNode, "MultiplicativeExpression", true);

        tokens.remove(0);

        double lastNextTwoArgExpressionChance = ((Expression) parentNode).getNEXT_TWO_ARG_EXPRESSION_CHANCE();
        leftExpression = new Expression(this, lastNextTwoArgExpressionChance, tokens);
        operator = new MultiplicativeOperator(this, tokens);
        rightExpression = new Expression(this, lastNextTwoArgExpressionChance, tokens);
    }
}

File: Interactive.Genetic.Programming.Environment/src/Domain/Model/Nodes.Small/Expressions.Standard/MultiplicativeExpression.cs
using Model.Abstract;
using Model.Interfaces;
using Model.Nodes.Small.Operators.Arithmetic;

namespace Model.Nodes.Small.Expressions.Standard;

public sealed class MultiplicativeExpression : TwoArgExpression, ICrossable, ISubtreeMutable
{
    public override StandardExpression Left { get; protected set; }
    public override Operator Operator { get; protected set; }
    public override StandardExpression Right { get; protected set; }
    
    public MultiplicativeExpression(Node parentNode) : base(parentNode, "MultiplicativeExpression", true)
    {
        var lastNextTwoArgExpressionChance = parentNode is StandardExpression expression
            ? expression.NextTwoArgExpressionChance
            : throw new InvalidOperationException("Parent node is not an expression");

        Left = new StandardExpression(this, lastNextTwoArgExpressionChance);
        Operator = new MultiplicativeOperator(this);
        Right = new StandardExpression(this, lastNextTwoArgExpressionChance);
    }

    public MultiplicativeExpression(Node parentNode, StandardExpression left, StandardExpression right, string op)
        : base(parentNode, "MultiplicativeExpression", true)
    {
        Left = left;

        Operator = new MultiplicativeOperator(this);
        while (Operator.Value == op)
        {
            Operator = new MultiplicativeOperator(this);
        }

        Right = right;
    }

    public MultiplicativeExpression(Node parentNode, List<Token> tokens) : base(parentNode, "MultiplicativeExpression", true)
    {
        tokens.RemoveAt(0);

        var lastNextTwoArgExpressionChance = parentNode is StandardExpression expression
            ? expression.NextTwoArgExpressionChance
            : throw new InvalidOperationException("Parent node is not an expression");

        Left = new StandardExpression(this, tokens, lastNextTwoArgExpressionChance);
        Operator = new MultiplicativeOperator(this, tokens);
        Right = new StandardExpression(this, tokens, lastNextTwoArgExpressionChance);
    }

    public override string ToString() => $"({Left} {Operator} {Right})";
}